File: Studio/Lab0/Proposta/consumatore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256
// consumatore.c e' un filtro
int main(int argc, char *argv[]) {
  char *file_in, *prefix;
  char err[MAX_STRING_LENGTH], buffer[MAX_STRING_LENGTH];

  int ret, found, nread, i, fd;
  char read_char;
  // controllo numero argomenti
  // se invoco con 2 argomenti utilizzero stdin come file di input
  // se invoco con 3 argomenti utilizzero il file passato come argomento
  if (argc < 2 || argc > 3) {
    printf("Error:%s prefix < filename \n OR \n %s prefix filename\n", argv[0],
           argv[0]);
    return EXIT_FAILURE;
  } else if (argc == 2) {
    printf("Invocation with only prefix\n");
    prefix = argv[1];
    // 0 Standard input (stdin)
    fd = 0;
  } else if (argc == 3) {
    printf("Invocation with prefix and file name\n");
    prefix = argv[1];
    file_in = argv[2];

    fd = open(file_in, O_RDONLY);
    if (fd < 0) {
      perror("P0: Impossibile aprire il file");
      return EXIT_FAILURE;
    }
  }
  /*
      file_in = argv[1];
      fd = open(file_in, O_RDONLY);
      if (fd < 0) {
          perror("P0: Impossibile aprire il file.");
          exit(2);
      }
  */

  found = 0;

  while ((nread = read(fd, &read_char, sizeof(char)))) /* Fino ad EOF*/ {
    if (nread < 0) {
      printf("(PID %d) impossibile leggere dal file %s", getpid(), file_in);
      perror("Errore in lettura");
      close(fd);
      exit(3);
      return EXIT_FAILURE;
    }

    // leggo il carattere e controllo se è presente nel prefix
    for (int i = 0; i < strlen(prefix); i++) {
      if (read_char == prefix[i]) {
        found = 1;
      }
    }

    if (!found) {
      printf("%c", read_char);
    }

    found = 0;
  }
  close(fd);
  return EXIT_SUCCESS;
}

File: Laboratori/Lab0/Proposta/Consumatore.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;

// Consumatore e' un filtro
public class Consumatore {
	public static void main(String[] args) {
		BufferedReader in = null;
		String filterString;
		int found, c;
		char ch;


		if (args.length < 1 || args.length>2){ //Invocation error
			System.out.println("Utilizzo: consumatore <prefix> [filename]");
			System.out.println("OR");
			System.out.println("Utilizzo: consumatore <prefix> < [filename]");
			System.exit(0);
		}

        if(args.length == 1){
            		System.out.println("Esecuzione con solo prefisso.");
			System.out.println("Prefisso = "+args[0] + " di lunghezza "+args[0].length());
			
             in = new BufferedReader(new InputStreamReader(System.in));
        } else if(args.length == 2){System.out.println("Esecuzione con prefisso e file.");
			System.out.println("Prefisso = "+args[0] + " di lunghezza "+args[0].length()+
					"\nFile Path = "+args[1]);
                    try {
                    in = new BufferedReader(new FileReader(args[1]));
                } catch(FileNotFoundException e){
                    System.out.println("File non trovato");
                    System.exit(1);
                }
        }
	  

		filterString = args[0];

		try {
            found = 0;
			while ((c = in.read()) > 0) { 
                ch = (char) c;
                for(int i = 0; i<filterString.length(); i++){
                    if(ch == filterString.charAt(i)){
                        found = 1;
                    }
                }
                if(found==0){
                    System.out.print(ch);
                }
                found=0;
			}
			in.close();
		} catch(IOException ex){
			System.out.println("Errore di input");
			System.exit(2);
		}
}}

// done